src/main/scala/edu/twitter/model/impl/neuralnetwork/DataIterator.scala

```diff
@@ -0,0 +1,196 @@
+    val positive = new ArrayBuffer[Boolean]()
+    var i = 0
+
+    while (i < num && dataCursor < totalExamples) {
+
+      val msg = dataList(dataCursor).getAs[String]("msg")
+      val label = dataList(dataCursor).getAs[Double]("label")
+
+      //TODO(elhawaty): Find why some rows return null.
+      if (msg != null) {
+        tweets.append(msg)
+        positive.append(if (label == 1.0) true else false)
+        i += 1
+      } else {
+        println("Invalid tweet:" + dataList(dataCursor))
+      }
+
+      dataCursor += 1
+    }
+
+    val allTokens = removeUnkownWords(tweets)
+    var maxLength = allTokens.maxBy(_.length).length
+
+    // Workaround, just in case the word2vec doesn't recognise all the words in the batch which is unlikely to happen as we are using Google word2vec.
+    if (maxLength == 0) {
+      println(tweets(0))
+      allTokens(0).append("times")
+      maxLength = 1
+    }
+
+    //If longest review exceeds 'truncateLength': only take the first 'truncateLength' words
+    if (maxLength > truncateLength) maxLength = truncateLength
+
+    transformToDataSet(allTokens, positive, maxLength)
+  }
+
+  /**
+    * Tokenize the tweets and filter out the unknown words
+    *
+    * @param tweets the tweets to be filtered
+    * @return the filtered tweets
+    */
+  private def removeUnkownWords(tweets: ArrayBuffer[String]): ArrayBuffer[mutable.Buffer[String]] = {
+    val allTokens = new ArrayBuffer[mutable.Buffer[String]](tweets.size)
+    import scala.collection.JavaConversions._
+    for (s <- tweets) {
+      val tokens = tokenizerFactory.create(s).getTokens
+      val tokensFiltered = tokens.filter(wordVectors.hasWord)
+      allTokens.append(tokensFiltered)
+    }
+    allTokens
+  }
+
+  /**
+    * Transform the tweet to a DataSet the contains the extracted Feature of the tweet and its label
+    *
+    * @param allTokens the filtered tweets
+    * @param positive  the label of a tweet
+    * @param maxLength the maximum length of a tweet
+    * @return DataSet contains both the features and the label of the tweets
+    */
+  private def transformToDataSet(allTokens: ArrayBuffer[mutable.Buffer[String]], positive: ArrayBuffer[Boolean], maxLength: Int): DataSet = {
+    val features = Nd4j.create(allTokens.size, vectorSize, maxLength)
+    val labels = Nd4j.create(allTokens.size, 2, maxLength)
+    //Two labels: positive or negative because we are dealing with reviews of different lengths and only one output at the final time step: use padding arrays mask arrays contain 1 if data is present at that time step for that example, or 0 if data is just padding
+    val featuresMask = Nd4j.zeros(allTokens.size, maxLength)
+    val labelsMask = Nd4j.zeros(allTokens.size, maxLength)
+
+    val temp = new Array[Int](2)
+    for (i <- allTokens.indices) {
+      val tokens = allTokens(i)
+      temp(0) = i
+      //Get word vectors for each word in review, and put them in the training data
+      var j: Int = 0
```

Comment:
`j` declared but never used

Reply:
Changed in my previous commit 'Responding to change requests made by 3mo Ramadan' 👍

Reply:
:+1: